### PR TITLE
Modernization and ambiguity reduction

### PR DESCRIPTION
**Pull request by CSharperMantle**
## Major changes
1. Use CMake as project making system.
2. Rewrite marco signatures in order to reduce ambiguity.
For example, `vector_set_capacity(vec, size)` is renamed to `cvector_set_capacity(vec, size)` to reduce the risk of clashing with other libraries.
Also, marco `LOGARITHMIC_GROWTH` is renamed to `CVECTOR_LOGARITHMIC_GROWTH` for the same reason.
3. `README.md` and `exmaple.c` are edited to suit the new source file.

### CODE REVIEW

File: cvector.h [hunk 1]
```diff
@@ -1,106 +1,115 @@
 
-#ifndef VECTOR_H_
-#define VECTOR_H_
+#ifndef __CVECTOR_H__
```
Comment (eteran) [inline cvector.h]: Identifiers with a double underscore or a leading underscore followed by capital letters are reserved by the standard. (I recognize I use a double underscore later, but let's not double down on violating the standard if we can.

File: cvector.c [hunk 2]
```diff
@@ -0,0 +1 @@
+#include "cvector.h"
```
Comment (eteran) [inline cvector.c]: Why add this C file? It's a header-only library.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.0.0)
+project(c-vector VERSION 0.1.0)
+
+add_library(c-vector cvector.c)
+add_executable (c-vector-example cvector.c example.c)
+
+set(CPACK_PROJECT_NAME ${PROJECT_NAME})
+set(CPACK_PROJECT_VERSION ${PROJECT_VERSION})
+include(CPack)
```
Comment (eteran) [inline CMakeLists.txt]: Do we really need CPack?

File: cvector.h [hunk 4]
```diff
@@ -1,106 +1,115 @@
...
 
 #include <stddef.h> /* for size_t */
 #include <stdlib.h> /* for malloc/realloc/free */
 #include <assert.h> /* for assert */
 
 /**
- * @brief vector_set_capacity - For internal use, sets the capacity variable of the vector
+ * @brief cvector_VECTOR - The vector type used in this library  
+ */
+#define cvector_VECTOR(type) type *
```
Comment (eteran) [inline cvector.h]: Not sure I **love** this new macro with the all caps suffix. Maybe `cvector_create` or something like that for consistency? I'm open to suggestion.

File: example.c [hunk 5]
```diff
@@ -1,59 +1,59 @@
...
+    size_t i;
+    for (i = 0; i < cvector_size(v); ++i) {
+      printf("v[%lu] = %d\n", i, v[i]);
+    }
+  }
+
+  /* well, we don't have destructors, so let's clean things up */
+  cvector_free(v);
+
+  return 0;
```
Comment (eteran) [inline example.c]: Why the switch away from tabs to 2 spaces? I could understand 4 spaces, as two is a little cramped for my taste. But the switch away from tabs in an example seems... pointless to me.

Also, I believe that ANSI C requires each source file to end with a newline.

Comment (eteran) [reply example.c]: NOTE: I do now notice that the `example.c` source file Had **strange** tabbing, I think I temporarily tabbed in it when preparing to copy into the `README.md`, and accidentally saved it. LOL

File: cvector.h [hunk 6]
```diff
@@ -1,106 +1,115 @@
 
-#ifndef VECTOR_H_
-#define VECTOR_H_
+#ifndef __CVECTOR_H__
+#define __CVECTOR_H__
+
+#ifdef __cplusplus
```
Comment (eteran) [inline cvector.h]: No need to wrap any of this in `extern "C"  blocks as it exports no symbols at all to worry about mangling. (The standard headers certainly already have this protection).

File: cvector.h (hunk 1 above)
Comment (CSharperMantle) [reply cvector.h]: Thanks for your correction. Yes, these double underscores are reserved. The have been corrected in the latest commit. I used to write a lot of them in my previous code and now I know that I have made mistakes. Thanks a lot bro :D

File: cvector.c (hunk 2 above)
Comment (CSharperMantle) [reply cvector.c]: Well I just thought that a project with only an example make target would be strange. This has been corrected in the latest commit.

File: CMakeLists.txt (hunk 3 above)
Comment (CSharperMantle) [reply CMakeLists.txt]: Actually we don't. These lines are default in the CMake extension of VSCode, along with the support of CTest. This has been corrected.

File: example.c (hunk 5 above)
Comment (CSharperMantle) [reply example.c]: The code was formatted by `clang-format` in a default configuration and I forgot to edit them (which is really sorry). Now the code is formatted correctly with 4 spaces as indentation. But in China we often use this kind of ugly but compact coding style when doing National Olympiad of Information. But I admit this is not a good practice when on Github! XD

File: cvector.h (hunk 6 above)
Comment (CSharperMantle) [reply cvector.h]: Yeah. This has been corrected.

**Comment by eteran** [comment]
Looking better, thanks!

Could we come up with a better name for the:

```
#define cvector_VECTOR(type) type *
```

macro though. It's not consistent with any of the other names in the header and I generally don't like lots of capitols.

**Comment by eteran** [comment]
@CSharperMantle Any progress on that last change. I'd be happy to merge the PR once that one last tweak is made :-)

**Comment by CSharperMantle** [comment]
@eteran Sorry for the long delay. I am sorry but there will be a final exam in China, and I have to get ready for it. Would it be possible for you to wait for another 5 days? Till then the exam will be over.

**Comment by eteran** [comment]
no problem at all. good luck on your exams!

**Comment by CSharperMantle** [comment]
Sorry for the long delay. I've changed the name as shown in the code, but I don't think I have any better ideas. Sorry for my lack of creativity.

**Comment by eteran** [comment]
Thanks.

Macros are definitely typically all caps on most coding styles. However, since the style of this header is all "function like", I feel that it looked better "snake cased".

**Pull request merged**